reference/normalize.go

```diff
@@ -0,0 +1,134 @@
+package reference
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/docker/distribution/digest"
+)
+
+var (
+	legacyDefaultDomain = "index.docker.io"
+	defaultDomain       = "docker.io"
+	defaultRepoPrefix   = "library/"
+	defaultTag          = "latest"
+)
+
+// NormalizedName parses a string into a named reference
+// transforming a familiar name from Docker UI to a fully
+// qualified reference. If the value may be an identifier
+// use ParseAnyReference.
+func NormalizedName(s string) (Named, error) {
+	if ok := anchoredIdentifierRegexp.MatchString(s); ok {
+		return nil, fmt.Errorf("invalid repository name (%s), cannot specify 64-byte hexadecimal strings", s)
+	}
+	domain, remainder := splitDockerDomain(s)
+	var remoteName string
+	if tagSep := strings.IndexRune(remainder, ':'); tagSep > -1 {
+		remoteName = remainder[:tagSep]
+	} else {
+		remoteName = remainder
+	}
+	if strings.ToLower(remoteName) != remoteName {
+		return nil, errors.New("invalid reference format: repository name must be lowercase")
+	}
+
+	return ParseNamed(domain + "/" + remainder)
+}
+
+// splitDockerDomain splits a repository name to domain and remotename string.
+// If no valid domain is found, the default domain is used. Repository name
+// needs to be already validated before.
+func splitDockerDomain(name string) (domain, remainder string) {
+	i := strings.IndexRune(name, '/')
+	if i == -1 || (!strings.ContainsAny(name[:i], ".:") && name[:i] != "localhost") {
+		domain, remainder = defaultDomain, name
+	} else {
+		domain, remainder = name[:i], name[i+1:]
+	}
+	if domain == legacyDefaultDomain {
+		domain = defaultDomain
+	}
+	if domain == defaultDomain && !strings.ContainsRune(remainder, '/') {
+		remainder = defaultRepoPrefix + remainder
+	}
+	return
+}
+
+// FamiliarName returns a shortened version of the name familiar
+// to to the Docker UI. Familiar names have the default domain
+// "docker.io" and "library/" repository prefix removed.
+// For example, "docker.io/library/redis" will have the familiar
+// name "redis" and "docker.io/dmcgowan/myapp" will be "dmcgowan/myapp".
+func FamiliarName(named Named) Named {
+	var repo repository
+	repo.domain, repo.path = splitDomain(named.Name())
+
+	if repo.domain == defaultDomain {
+		repo.domain = ""
+		if strings.HasPrefix(repo.path, defaultRepoPrefix) {
+			repo.path = strings.TrimPrefix(repo.path, defaultRepoPrefix)
+		}
+	}
+	if digested, ok := named.(Digested); ok {
+		return canonicalReference{
+			repository: repo,
+			digest:     digested.Digest(),
+		}
+	}
+	if tagged, ok := named.(Tagged); ok {
+		return taggedReference{
+			repository: repo,
+			tag:        tagged.Tag(),
+		}
+	}
+	return repo
+}
+
+// WithDefaultTag adds the default tag "latest" to a reference if it only has
+// a repo name.
+func WithDefaultTag(ref Named) NamedTagged {
```

Comment:
I know this is adapted from docker/docker, but I find the function confusing. Based on the name, I would expect it to override an existing tag with the default tag. Can we call it `EnsureTagged` or something like that?